### PR TITLE
[Merged by Bors] - Initial version of a JS -> Rust conversion trait.

### LINKED ISSUE #1975: TryFromJsValue
Working on [Azle](https://github.com/demergent-labs/azle) which is a JS environment for the Internet Computer, I need to have a general solution for converting `JsValue`s into arbitrary types. I have been attempting to do this using `serde_json::from_value(js_value.to_json(& mut context).unwrap()).unwrap()` which works in many cases, but breaks down sometimes (for example with `undefined` and `bigint`). It's also a more complicated developer experience and probably less performant than a direct conversion from `JsValue` to an arbitrary type would be.

The below is from this comment: https://github.com/boa-dev/boa/issues/1961#issuecomment-1075673692

> This would make a lot of sense, yes. We could have a TryFromJsValue trait that would receive a context and a JsValue and try to perform the conversion. Just thinking out loud, but I could imagine a similar API to the Serializa/Deserialize APIs in serde.

I would like to start a discussion on some sort of `TryFromJsValue`. Is this something you think boa could/would want to support? If so, what should the API look like?

I would like this functionality ASAP and so if it would be accepted and I could receive some guidance, I am happy to do the work. If it's relatively simple for the maintainers to go ahead and do it then I am happy with that as well. Just hope this can be done.

Thanks!

### PR DESCRIPTION
**Pull request by Razican**
This Pull Request closes #1975. It's still a work in progress, but tries to go in that direction.

It changes the following:

- Adds a new `TryFromJs` trait, that can be derived using a new `boa_derive` crate.
- Adds a new `try_js_into()` function that, similarly to the standard library `TryInto` trait

Things to think about:
- Should the `boa_derive` crate be re-exported in `boa_engine` using a `derive` feature, similar to how it's done in `serde`?
- The current implementation only converts perfectly valid values. So, if we try to convert a big integer into an `i8`, or any floating point number to an `f32`. So, you cannot derive `TryFromJs` for structures that contain an `f32` for example (you can still manually implement the trait, though, and decide in favour of a loss of precision). Should we also provide some traits for transparent loss of precision?
- Currently, you cannot convert between types, so if the JS struct has an integer, you cannot cast it to a boolean, for example. Should we provide a `TryConvertJs` trait, for example to force conversions?
- Currently we only have basic types and object conversions. Should add `Array` to `Vec` conversion, for example, right? Should we also add `TypedArray` conversions? What about `Map` and `Set`? Does this step over the fine grained APIs that we were creating?

Note that this still requires a bunch of documentation, tests, and validation from the dev team and from the users that requested this feature. I'm particularly interested in @lastmjs's thoughts on this API.

I already added an usage example in `boa_examples/src/bin/derive.rs`.

### CODE REVIEW

**Comment by Razican** [comment]
After some thoughts coming from the feedback of our users, this would need an extra attribute for each member, so that users can override the transformation function for a particular field. Similarly to `deserialize_with` in `serde`. I'll try to implement that next when I have some time.

**Comment by Razican** [comment]
I have added conversion overrides with attributes in the derive. This should make it easy to implement specific conversions for types that either do not implement `TryFromJs` or for types where we want to be a bit more flexible.

The example in [`boa_examples/src/bin/derive.rs`](https://github.com/boa-dev/boa/blob/try_from_js/boa_examples/src/bin/derive.rs) uses this override for one of the fields.

Since this was a feature requested by @lastmjs, I would like to get some feedback and see if he likes this approach.

**Comment by lastmjs** [comment]
What if we want to override these traits for the primitive types, like `u128`? The solution we've moved to is just implementing our own traits, `TryIntoVmValue` and `TryFromVmValue`. We implement the trait on all of our types, and in the implementations we either use the builtin boa conversions, or if necessary our own custom conversion.

**Comment by jedel1043** [comment]
> What if we want to override these traits for the primitive types, like `u128`? The solution we've moved to is just implementing our own traits, `TryIntoVmValue` and `TryFromVmValue`. We implement the trait on all of our types, and in the implementations we either use the builtin boa conversions, or if necessary our own custom conversion.

If just a few types will be overridden, couldn't you just use an intermediary newtype, and implement `TryFromJs`/`TryIntoJs` for it? Or is it too intrusive for your specific API?

**Comment by Razican** [comment]
I'm setting this as "ready for review", since there are a few things that can be implemented, but, I don't have the needed time right now, and this is already useful as-is. Things for the future:

 - Conversion back from Rust to JS (which currently would require to use a native object or the `Into` trait)
 - Conversions of `TypedArray`. I tried to implement this, but it's not so simple to get the buffer and convert it to the proper vector in Rust. This also requires some decisions:
    - Do we want to clone the buffer? (we are doing this for `BigInt`, btw
    - Do we want to "take" the buffer and leave the JS object without it? This would mean that subsequent calls inside JavaScript would get different results.

I think this is the minimal implementation that we can provide without making sacrifices in the long run.

**Comment by Razican** [comment]
I think this is ready for review. Could we merge it as-is?

In the future, we could use this trait to convert `Vec<u8>` and others from `ArrayBuffer` or `TypedArray`. It would be nice to also have a similar trait for the Rust to JS conversions, but I guess this can already be reviewed and merged.

**Comment by raskad** [comment]
Does it make sense to put the macros in `boa_macros` instead of adding the new crate? In `boa_engine` we already depend on `boa_macros` anyway.

**Comment by Razican** [comment]
> Does it make sense to put the macros in `boa_macros` instead of adding the new crate? In `boa_engine` we already depend on `boa_macros` anyway.

Indeed, this was started when `boa_macros` didn't exist. I have now moved it into `boa_macros`, and re-exported the derive in `value::TryFromJs`, so now, importing the trait automatically imports the derive too.

**Comment by Razican** [comment]
I'm unable to reproduce the tarpaulin error locally :/

**Comment by raskad** [comment]
> I'm unable to reproduce the tarpaulin error locally :/

Me neither. My only guess is that it may be caused by the cyclic dependency between `boa_macros` and `boa_engine`. I would suggest we test if it works when we remove the `boa_engine` dev-dependency from `boa_macros`.

**Comment by nekevss** [comment]
Seconding the cyclical dependency guess raskad mentioned above. I ran into something similar with cyclical dependencies when adding `Trace` and `Finalize`.

Would it be worthwhile to move the derive macro into a `derive_macro` folder under `boa_engine` and then re-export the value for a public use in `boa_macros`? I think Bevy does something similar to this where crates own their derive macros.

**Comment by jedel1043** [comment]
I'm pretty sure derive crates don't test their implementations internally. For example, `serde` has a `serde_test` crate on its workspace precisely for this reason. If we want to test, we can either:

- Create a `boa_test` crate for integration tests (We could also move all integration tests there, in order to speedup compilation times).

- Move the `boa_derive` tests to `boa_engine`.

I'm tending towards the first option, because we have a LOT of integration tests in a single, massive file and we could benefit from an exclusive crate for tests.

**Comment by Razican** [comment]
> I'm pretty sure derive crates don't test their implementations internally. For example, `serde` has a `serde_test` crate on its workspace precisely for this reason. If we want to test, we can either:
> 
> - Create a `boa_test` crate for integration tests (We could also move all integration tests there, in order to speedup compilation times).
> 
> - Move the `boa_derive` tests to `boa_engine`.
> 
> I'm tending towards the first option, because we have a LOT of integration tests in a single, massive file and we could benefit from an exclusive crate for tests.

I guess this is the best option. I would say that for now, I can just move these tests to `boa_engine`, and then we can create another PR to add all the tests into a crate.

**Comment by Razican** [comment]
Rebased :) feel free to review it and see if we can already merge it as-is (and improve it later), or if you'd like me to add some more stuff

File: boa_engine/src/value/conversions/mod.rs [hunk 1]
```diff
@@ -1,5 +1,19 @@
...
+pub(super) mod try_from_js;
+
+impl From<&Self> for JsValue {
+    #[inline]
+    fn from(value: &Self) -> Self {
+        let _timer = Profiler::global().start_event("From<&JsValue>", "value");
+
+        value.clone()
+    }
+}
```
[the remark this answers is not in the document]
Comment (raskad) [reply boa_engine/src/value/conversions/mod.rs:15]: I think let's keep it removed.

**Review by raskad — APPROVED** [review]
Really cool feature. The example is also very nice.

**Review by HalidOdat — APPROVED** [review]
Looks perfect to me! :)

**Comment by Razican** [comment]
It seems that tarpaulin is failing somehow, but I cannot reproduce it in a Mac

**Comment by Razican** [comment]
I have rebased this and updated the code for syn 2.0, which is now much cleaner :)

**Comment by Razican** [comment]
@HalidOdat, @raskad, @jedel1043, @RageKnify, @nekevss, @jasonwilliams did you have the chance to run `cargo tarpaulin` locally to see if you can reproduce the error?

```
error: Broken pipe (os error 32)
warning: build failed, waiting for other jobs to finish...
error: could not compile `boa_cli` due to 3 previous errors
Mar 25 19:08:10.059 ERROR cargo_tarpaulin: Failed to compile tests!
error[E0308]: mismatched types
   --> boa_cli/src/main.rs:196:29
    |
196 |     boa_parser::Parser::new(Source::from_bytes(&src))
    |     ----------------------- ^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `boa_parser::Source`, found struct `boa_engine::Source`
    |     |
    |     arguments to this function are incorrect
    |
    = note: struct `boa_engine::Source` and struct `boa_parser::Source` have similar names, but are actually distinct types
note: struct `boa_engine::Source` is defined in crate `boa_parser`
   --> /home/runner/work/boa/boa/boa_parser/src/source.rs:12:1
    |
12  | pub struct Source<'path, R> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: struct `boa_parser::Source` is defined in crate `boa_parser`
   --> /home/runner/work/boa/boa/boa_parser/src/source.rs:12:1
    |
12  | pub struct Source<'path, R> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: perhaps two different versions of crate `boa_parser` are being used?
note: associated function defined here
   --> /home/runner/work/boa/boa/boa_parser/src/parser/mod.rs:121:12
    |
121 |     pub fn new(source: Source<'a, R>) -> Self {
    |            ^^^


Error: "Failed to compile tests!\nerror[E0308]: mismatched types\n   --> boa_cli/src/main.rs:196:29\n    |\n196 |     boa_parser::Parser::new(Source::from_bytes(&src))\n    |     ----------------------- ^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `boa_parser::Source`, found struct `boa_engine::Source`\n    |     |\n    |     arguments to this function are incorrect\n    |\n    = note: struct `boa_engine::Source` and struct `boa_parser::Source` have similar names, but are actually distinct types\nnote: struct `boa_engine::Source` is defined in crate `boa_parser`\n   --> /home/runner/work/boa/boa/boa_parser/src/source.rs:12:1\n    |\n12  | pub struct Source<'path, R> {\n    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^\nnote: struct `boa_parser::Source` is defined in crate `boa_parser`\n   --> /home/runner/work/boa/boa/boa_parser/src/source.rs:12:1\n    |\n12  | pub struct Source<'path, R> {\n    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n    = note: perhaps two different versions of crate `boa_parser` are being used?\nnote: associated function defined here\n   --> /home/runner/work/boa/boa/boa_parser/src/parser/mod.rs:121:12\n    |\n121 |     pub fn new(source: Source<'a, R>) -> Self {\n    |            ^^^\n\n"
```

**Comment by jedel1043** [comment]
I cannot reproduce it from my machine. I think it's probably a caching issue.

**Comment by jedel1043** [comment]
Let's see if this works...

**Comment by jedel1043** [comment]
So neither the cache nor the action are the cause...

**Comment by jedel1043** [comment]
Huh, removing the cyclic test dep fixed it...
I see two possible explanations:
- A Tarpaulin bug on certain environments with cyclic dependencies
- Removing the tests triggered a recompilation

Was about to check if re adding the tests fixed the error but I had to go out, so feel free to uncomment the commented out code, otherwise I'll do that in a couple of hours.

**Comment by Razican** [comment]
I'll try to move the tests to a new crate.

**Comment by Razican** [comment]
It seems this solves the issue :) Should we merge this?

**Comment by jedel1043** [comment]
> It seems this solves the issue :) Should we merge this?

Sounds good!

bors r+

**Review by raskad — APPROVED** [review]

**Pull request closed without merging**